### PR TITLE
Store download directory in database

### PR DESCRIPTION
**Pull request by Lipt0nas**
Also split up the started events into separate ones for downloading and uploading to make it a bit cleaner

### CODE REVIEW

File: drop-storage/migrations/20230406110818_initial_migration.sql [hunk 1]
```diff
@@ -113,6 +113,7 @@ CREATE TABLE IF NOT EXISTS incoming_path_pending_states (
 CREATE TABLE IF NOT EXISTS incoming_path_started_states (
   id INTEGER PRIMARY KEY AUTOINCREMENT, 
   path_id INTEGER NOT NULL,
+  base_dir TEXT NOT NULL,
```
Comment (Lipt0nas) [inline drop-storage/migrations/20230406110818_initial_migration.sql:116]: Not sure if this is the optimal naming, suggestions are welcome

Comment (matszczygiel) [reply drop-storage/migrations/20230406110818_initial_migration.sql:117]: I would go for `dest_dir` but it's a soft suggestion as it's fine already

File: drop-storage/src/lib.rs [hunk 2]
```diff
@@ -293,16 +293,18 @@ impl Storage {
         &self,
         transfer_id: Uuid,
         path_id: String,
+        base_dir: String,
```
Comment (matszczygiel) [inline drop-storage/src/lib.rs:296]: I'm not sure why we take the owned string in the DB methods, `&str` should suffice

File: drop-transfer/src/storage_dispatch.rs [hunk 3]
```diff
@@ -40,22 +40,24 @@ impl<'a> StorageDispatch<'a> {
...
+
+            Event::FileDownloadStarted {
+                transfer_id,
+                file_id,
+                base_dir,
+            } => {
+                self.storage
+                    .insert_incoming_path_started_state(transfer_id, file_id, base_dir)
+                    .await?
+            }
```
Comment (matszczygiel) [inline drop-transfer/src/storage_dispatch.rs:60]: Nice :)

File: drop-transfer/src/storage_dispatch.rs [hunk 4]
```diff
@@ -178,13 +180,14 @@ impl From<&crate::Event> for Event {
...
-                transfer_id: transfer.id(),
-                file_id: file.to_string(),
-            },
-            crate::Event::FileUploadStarted(transfer, file) => Event::Started {
-                transfer_type: TransferType::Outgoing,
+            crate::Event::FileDownloadStarted(transfer, file, base_dir) => {
+                Event::FileDownloadStarted {
+                    transfer_id: transfer.id(),
+                    file_id: file.to_string(),
+                    base_dir: base_dir.clone(),
```
Comment (matszczygiel) [inline drop-transfer/src/storage_dispatch.rs:187]: `
```suggestion
                    base_dir: base_dir,
```

File: drop-transfer/src/ws/server/mod.rs [hunk 5]
```diff
@@ -412,11 +413,17 @@ async fn handle_client(
 }
 
 impl FileXferTask {
-    pub fn new(file: crate::File, xfer: crate::Transfer, location: PathBuf) -> crate::Result<Self> {
+    pub fn new(
+        file: crate::File,
+        xfer: crate::Transfer,
+        location: PathBuf,
+        base_dir: PathBuf,
```
Comment (matszczygiel) [inline drop-transfer/src/ws/server/mod.rs:420]: Isn't the location the same thing as base_dir ?

Comment (Lipt0nas) [reply drop-transfer/src/ws/server/mod.rs:420]: location here is the already glued together final path of the file

File: drop-transfer/src/storage_dispatch.rs (hunk 4 above)
Comment (Lipt0nas) [reply drop-transfer/src/storage_dispatch.rs:187]: `base_dir` from the event is `&String`, can't not clone here

File: drop-storage/src/types.rs [hunk 6]
```diff
@@ -49,10 +49,14 @@ pub enum Event {
     Pending {
         transfer_info: TransferInfo,
     },
-    Started {
-        transfer_type: TransferType,
+    FileUploadStarted {
+        transfer_id: TransferId,
+        file_id: FileId,
+    },
+    FileDownloadStarted {
```
Comment (LukasPukenis) [inline drop-storage/src/types.rs:56]: :chef's kiss:

File: drop-transfer/src/ws/server/mod.rs (hunk 5 above)
Comment (LukasPukenis) [reply drop-transfer/src/ws/server/mod.rs:420]: `full_location` or `absolute_path` are better names to describe `location`

**Review by LukasPukenis — DISMISSED** [review]
There is some confusion around the naming of the fields that could be improved but that's not a blocker

File: drop-transfer/src/storage_dispatch.rs (hunk 4 above)
Comment (matszczygiel) [reply drop-transfer/src/storage_dispatch.rs:187]: No, if it's `&String` then it's fine. I thought it was `String`

File: drop-transfer/src/ws/server/mod.rs (hunk 5 above)
Comment (matszczygiel) [reply drop-transfer/src/ws/server/mod.rs:420]: Yes, the naming could be changed for a more descriptive one

Comment (matszczygiel) [reply drop-transfer/src/ws/server/mod.rs:420]: TBH we can get rid of `location` here as we can construct location as `base_dir` + `file.subpath()`

Comment (Lipt0nas) [reply drop-transfer/src/ws/server/mod.rs]: I agree with the naming, but getting rid of the location IMO might be a bit too of an intrusive change, as we don't trivially concatenate the paths, but also perform mapping and validation on it as well, so doing it once and storing, like we are doing now seems like a better solution

**Pull request merged**